argon2_out_of_process_support.go

```diff
@@ -0,0 +1,990 @@
+// returned from go routines that are created during the handling of a request may be returned directly
+// from this function to be handled by the current process.
+//
+// Note that this function won't return until the supplied io.Reader is closed by the parent.
+func WaitForAndRunArgon2OutOfProcessRequest(in io.Reader, out io.Writer, watchdog Argon2OutOfProcessWatchdogHandler) error {
+	if atomic.LoadUint32(&argon2OutOfProcessHandlerStatus) == inProcessArgon2KDFUsed {
+		return ErrArgon2OutOfProcessHandlerUnavailable
+	}
+
+	if watchdog == nil {
+		watchdog = NoArgon2OutOfProcessWatchdogHandler()
+	}
+
+	tmb := new(tomb.Tomb)
+
+	// rspChan is the channel from the routines that process requests and run the KDF to the
+	// dedicated output routine which serializes the response to the supplied io.Writer.
+	rspChan := make(chan *Argon2OutOfProcessResponse)
+
+	// Spin up a routine for receiving requests from the supplied io.Reader.
+	tmb.Go(func() error {
+		// Also spin-up the routine for sending outgoing responses that are generated internally.
+		// This handles the read end of rspChan, and serializes responses to the supplied io.Writer.
+		// This gets its own goroutine so that all responses are sent via the same code path - responses
+		// can ultimately come directly from the request processing loop in this routine (in the event
+		// of a watchdog request), or from a dedicated KDF routine which permits the request processing
+		// loop in this routine to continue executing whilst the KDF is running, so we can continue to
+		// process watchdog requests.
+		tmb.Go(func() error {
+			// Loop whilst the tomb is alive.
+			for tmb.Alive() {
+				// Wait for a response from somewhere or wait for the tomb to
+				// begin dying.
+				select {
+				case rsp := <-rspChan:
+					// We've got a response from somewhere. Encode it send the
+					// response out on the io.Writer. If this fails, return an error,
+					// which begins the dying of this tomb and will result in an error
+					// being returned to the caller.
+					enc := json.NewEncoder(out)
+					if err := enc.Encode(rsp); err != nil {
+						return fmt.Errorf("cannot encode response: %w", err)
+					}
+				case <-tmb.Dying():
+					// We've begun to die, and this loop will not run again.
+				}
+			}
+			return tomb.ErrDying
+		})
+
+		// Run a loop for receiving and processing incoming requests from the io.Reader as
+		// long as the tomb remains alive.
+		for tmb.Alive() {
+			// Wait for a request from the io.Reader. The only way to unblock this is
+			// if the parent sends something or closes its end of the OS pipe. If it's
+			// closed before we've received a KDF request, then we terminate this routine
+			// with an error to begin the process of the tomb dying and the function
+			// eventually returning an error to the caller.
+			var req *Argon2OutOfProcessRequest
+			dec := json.NewDecoder(in)
+			dec.DisallowUnknownFields()
```

Comment:
means that the both ends need to be in sync now, which hopefully is true most of the time, but should there be some `format`, `version` field in requests/responses to indicate protocol version?

Just to be clear, I would not want to introduce any form to a handshake, as this would complicate the code even more, but a clear version check would help with debugging.

Reply:
I've removed the calls to `DisallowUnknownFields` everywhere it was used.